Web responsive menu improvements

danieltorres7 opened the pull request:
Added apptitle before oe_submenu
Increased user avatar size to maximum
Fullscren dropdown menus on mobile mode

pedrobaeza commented:
There are conflicts in the branch

danieltorres7 commented:
@pedrobaeza, all conflicts solved.

pedrobaeza commented on web_responsive/static/src/less/main.less:
```diff
@@ -45,6 +45,15 @@ main {
+// Remove z-index to preserve menu
+.o_control_panel {
+    .btn {
+        z-index: initial!important;
```
Please don't use important, as it overwrites everything else. Use a more specific selector instead.

pedrobaeza commented:
In general, you must avoid to use `!important` in CSS styling

danieltorres7 commented:
@pedrobaeza. Succesfully replaced all important decorators by more specific selectors.

danieltorres7 commented:
Check it out @rousseldenis, @lasley and @leorochael. Everything seems to be working with the new menus.

pedrobaeza reviewed (approved):
Tested on runbot

pedrobaeza commented:
@Yajo, can you review technically?

danieltorres7 committed:
Bold titles for apps in the app-drawer

danieltorres7 committed:
Remove z-index from Control Panel buttons

So it doesn't overlap the menu.

danieltorres7 committed:
Better responsive menus

Move systray icons/menus out of the menu and up to the top bar.

Increase avatar size and visibility.

Other minor menu behavior fixes.

danieltorres7 commented:
I've completely rewritten the commits, removing redundant stuff and adding better comments and commit messages, but the behavior should remain the same (save for a small animation bug that I fixed).

lasley commented on web_responsive/static/src/less/navbar.less:
```diff
@@ -18,12 +18,159 @@
+        // Navbar header mobile styling
+        margin-left: -11px;
+        margin-top: 5px;
+        margin-bottom: 5px;
```
Should we maybe use the Bootstrap vars here for consistency? - https://github.com/odoo/odoo/blob/10.0/addons/web/static/lib/bootstrap/less/variables.less#L361

yajo commented on web_responsive/static/src/less/navbar.less:
```diff
@@ -18,12 +18,159 @@
+        > li.dropdown-header {
+            color: white;
+            text-decoration: none;
+            background-color: #8E8E8E;
```
I think you'd better use the corresponding [Odoo](https://github.com/odoo/odoo/blob/10.0/addons/web/static/src/less/variables.less) or [Bootstrap](https://github.com/odoo/odoo/blob/10.0/addons/web/static/lib/bootstrap/less/variables.less) variables for color and background-color.

yajo commented on web_responsive/static/src/less/navbar.less:
```diff
@@ -18,12 +18,159 @@
+    }
+
+    // Fitt's law: Stick menu to the top of the screen
+    @media (min-width: @screen-sm) {
```
`@screen-sm` is deprecated. Use `@screen-sm-min`.

yajo commented on web_responsive/static/src/less/navbar.less:
```diff
@@ -18,12 +18,159 @@
+        }
+
+        // Fitt's law: Stick app-drawer button to the top-left of the screen
+        @media (min-width: @screen-sm) {
```
Deprecated.

yajo commented on web_responsive/static/src/less/navbar.less:
```diff
@@ -18,12 +18,159 @@
+        }
+    }
+
+    @media (min-width: @screen-sm) {
```
Deprecated.

yajo commented on web_responsive/static/src/less/navbar.less:
```diff
@@ -18,12 +18,159 @@
+        margin: 0px;
+        margin-right: -16px;
+
+        @media (min-width: @screen-sm) {
```
Deprecated.

yajo commented on web_responsive/static/src/less/navbar.less:
```diff
@@ -18,12 +18,159 @@
+                overflow: auto;
+                float: left;
+                margin-top: 2px;
+                background-color: #ffffff;
```
Again does any available variable fit in?

yajo reviewed (changes requested):
Besides some technical details below, I think the top bar should be thick in sm size like it is in xs size, to make it more usable in tablets.

danieltorres7 committed:
Changes to fit odoo variables

And a little margin adjust on navbar-right

danieltorres7 commented:
@Yajo and @pedrobaeza. Requested changes implemented.

danieltorres7 commented:
> Besides some technical details below, I think the top bar should be thick in sm size like it is in xs size, to make it more usable in tablets.

@Yajo, I agree on you, anyway, how can we make thick top bars for tablets not changing desktop ones? I mean, resolution is higher than 768px on both and there is not available decorators (at least I don't know at the moment) to easily check if the device is a tablet.

yajo commented:
There must be a media query that makes the bar thick in xs screens (or thin in >=sm). All you have to do is find it and change the width variable from xs to sm.

Oh, and about tablet or small PC, as long as tablet UI is usable in a small PC, you shouldn't care about desktop. At least until the [pointer media query](https://developer.mozilla.org/en/docs/Web/CSS/@media/pointer) lands mainstream.

danieltorres7 commented:
Thick top menus are already available for all sm screens, not only xs.

leorochael commented:
@Yajo, do you want the top menubar to remain thick, no matter how large the screen?

This way the only difference between larger and smaller screens would be the presence of the menus directly on the top bar, or inside the menu button.

Is this what you want?

danieltorres7 commented:
I will refactor changes, so top menu height will remain the same on all modes, is that what you want? Always thick no matter what mode is active?

pedrobaeza commented:
He will answer tomorrow, as he works only up to 15:00 CET

lasley commented:
> Always thick no matter what mode is active?

@danieltorres7 @leorochael - I think he is saying always thick when on XS and SM, but leave it for the others. I think I agree, that would be a great change.

rousseldenis reviewed (approved):
:+1: Tested on runbot

lasley commented on web_responsive/static/src/less/navbar.less:
```diff
@@ -18,16 +18,166 @@
     
+    // App name before oe_submenu
+    .app-name {
+        > .oe_menu_text {
```
Somewhat nitpicky, but maybe a `cursor: default` here to override `text`

lasley reviewed (approved):
@danieltorres7 Excellent improvement, few comments but otherwise 👍 

Looking at this, the navbar changing @Yajo is referring to is controlled with the [navbar-collapse, navbar-toggle](https://github.com/OCA/web/blob/10.0/web_responsive/views/web.xml#L43) classes.

I think maybe we should add this one to the roadmap. It's going to be getting into changing the way Bootstrap works, so I think it'll cause some cascading changes that'll blow this PR into a rabbit hole.

@Yajo - correct me if I'm wrong on my inferences.

yajo reviewed (dismissed):
I'm afrraid not, friend. Just open the runbot and resize the window to sm size (768-991px width) and you'll see the top bar is thinner than xs (<=767px).

In any case, I'm not blocking this for that reason, but if you want to push that, it would be great 😉 

Note to PSC: **squash** at merge.

yajo reviewed (approved):
OK now I have been able to review properly. Indeed I meant what @lasley says. This looks good to me, but I'll wait for @danieltorres7 feedback before merging: do you want to address those changes @lasley said before we merge?

And about the thick bar, if it is too complicated, then maybe just a roadmap note would work for now.

Thanks!

leorochael commented:
Ok, but considering that we can't tell large tablets apart from desktops before the pointer media query lands, wouldn't it be better (and simpler in the code) to just keep a wide top bar, no matter the resolution?

That's what Odoo does by default, BTW

danieltorres7 committed:
Top menu always thick

danieltorres7 commented:
Top menu now is always thick, simplified some styles for that.

yajo commented on web_responsive/static/src/less/main.less:
```diff
@@ -4,6 +4,7 @@
 body {
     width: 100%;
     height: 100%;
+    overflow: hidden!important; // Important is to override base template
```
You knew this was going to happen... Please change this 😆 

A simple trick for these kind of overriding is:

```
html & {
  [syles...]
}
```

`html body` is more concise than `body` selector, so it will win the fight 😉

leorochael replied:
What @danieltorres7 mean by "base template", is that there's a `style="overflow: auto;"` in the `<body>` tag ` *directly in the DOM*.

There's no way to add a more specific selector without `!important`

leorochael replied:
Apparently this `style` attribute is being added by The `drawer.css` library because `<body>` has a `drawer` class, added by the `web_responsive` module. We're still investigating.

leorochael replied:
This happens on the `drawer close` event of the app drawer, which happens automatically on the closing of the app drawer, or on window resize (triggers the drawer close).

At this point we would help suggestions on how to deal with the issue and how to remove the `!important`.

danieltorres7 replied:
Just removed overflow: hidden!important from main.less. Library drawer.3.2.0.js adds "overflow: auto" on close, also triggered on resize like @leorochael told us. This causes body to have it's own scrollbar and at small height screens we got 2 scrollbars at the same time. Committing that change on lib separately now...

yajo reviewed (changes requested):
You know what? It looks awesome. Just one more little change to merge 🎉

danieltorres7 committed:
Removed overflow:hidden!important from main.less

danieltorres7 committed:
Local patch: prevent body overflow auto

leorochael commented on web_responsive/static/lib/js/drawer.3.2.0.js:
```diff
@@ -136,7 +136,11 @@
         .removeClass(__.settings.class.open)
         .addClass(__.settings.class.close)
-        .css({ 'overflow': 'auto' })
+        // TODO: remove this local patch once PR below is approved:
```
Given [@Yajo's comment on 511](https://github.com/OCA/web/pull/511#issuecomment-269595936), This line should probably be just removed, and the following block added to `web_responsive/static/lib/css/drawer.3.2.0.css`:

    .drawer-open {
        overlay: hidden;
    }

After doing the same in a PR for drawer, of course (which means, I should probably [update my own PR](https://github.com/blivesta/drawer/pull/36)...)

danieltorres7 committed:
Fix of chat window z-index

Also centered vertically app-name

leorochael committed:
Stop using `style=` on `<body>`

Fix jquery drawer in anticipation of:
https://github.com/blivesta/drawer/pull/36

danieltorres7 committed:
Merge pull request #1 from idealcare/web_responsive_menu_improvements

Stop using `style=` on `<body>`

danieltorres7 commented:
@Yajo, requested changes and final fixes applied. We are ready for the final review.

leorochael commented:
Please, let us know when it's ok to squash/reorder the commits.

lasley commented:
@Yajo - the changes look good to you? 

@leorochael - Once he confirms, you're good to squash/reorder.

rousseldenis commented:
Come back from holidays and changes seem great !
Thanks for the job !

yajo commented:
Please squash to merge 😉

pedrobaeza commented:
I have squashed on merge.

pedrobaeza commented:
Can you please backport this to v9 or isn't direct?
